### PR TITLE
add to_paifu()

### PR DESCRIPTION
**Pull request by canuse**
#12 and #20 
Add to_paifu() to TenhouRecord, TenhouGame and TenhouEvent.
As one TenhouEvent may result in multiple lines (like INIT event), the return value of TenhouEvent.to_paifu() is a list containing the dict(s).

For some fields can not be inferred from TenhouEvent (like the round, subround and subsubround), TenhouRecord.to_paifu() and TenhouGame.to_paifu() will contain more details. 
So TenhouRecord.to_paifu() is recommended.
The game uuid, game info, player names, however, is not parsed in these functions.

### CODE REVIEW

File: mahjong/record/utils/event.py [hunk 1]
```diff
@@ -112,3 +113,99 @@ def __str__(self):
...
+                         'player_show': ''.join([str(tile_from_tenhou(int(i))) for i in self_tile]),
+                         'player_open': ''.join([str(tile_from_tenhou(int(i))) for i in borrow_tile]) + ''.join(
+                             [str(tile_from_tenhou(int(i))) for i in self_tile]), 'origin': source}]
+            elif type_of.chakan:
+                return [{'event_type': 'GANG', 'player': player,
+                         'player_show': ''.join([str(tile_from_tenhou(int(i))) for i in self_tile]),
+                         'player_open': ''.join([str(tile_from_tenhou(int(i))) for i in borrow_tile]) + ''.join(
+                             [str(tile_from_tenhou(int(i))) for i in self_tile]), 'origin': source}]
+            elif type_of.nuki:
+                return [{'event_type': 'KITA ', 'player': player,
```
Comment (Ledenel) [inline mahjong/record/utils/event.py]: `'KITA '` leaves a blanke

File: mahjong/record/utils/event.py [hunk 2]
```diff
@@ -112,3 +113,99 @@ def __str__(self):
...
+        if is_open_hand(event):
+
+            player = attrs['who']
+            item = meld_from(event)
+            type_of = unpack_with(MeldTypeData, meld_type_unpacker, event.attrib['m'])
+            source = item.from_who
+            self_tile = item.self_tiles
+            borrow_tile = item.borrowed_tiles
+            if type_of.syuntsu:
+                return [{'event_type': 'CHI  ', 'player': player,
```
Comment (Ledenel) [inline mahjong/record/utils/event.py]: `'CHI  '`remains two blanks. Is this meanful?

File: mahjong/record/utils/event.py [hunk 3]
```diff
@@ -112,3 +113,99 @@ def __str__(self):
...
+                         'player_show': ''.join([str(tile_from_tenhou(int(i))) for i in self_tile]),
+                         'player_open': ''.join([str(tile_from_tenhou(int(i))) for i in borrow_tile]) + ''.join(
+                             [str(tile_from_tenhou(int(i))) for i in self_tile]), 'origin': source}]
+            elif type_of.koutsu:
+                return [{'event_type': 'PENG ', 'player': player,
```
Comment (Ledenel) [inline mahjong/record/utils/event.py]: `'PENG '` remains a blank. Is this meanful?

File: mahjong/record/utils/event.py [hunk 4]
```diff
@@ -112,3 +113,99 @@ def __str__(self):
...
+                dict3 = {'event_type': 'INIT4', 'player': '3', 'player_see': ''.join(
+                    [str(tile_from_tenhou(i)) for i in sorted([int(i) for i in attrs['hai3'].split(',')])]),
+                         'score': attrs['ten'].split(',')[3] + '00'}
+                return [dict0, dict1, dict2, dict3, dora, initround]
+            else:
+                return [dict0, dict1, dict2, dora, initround]
+        draw = draw_tile_change(event)
+        if draw:
+            player, tile = draw['player'], draw['tile']
+            return [{'event_type': 'DRAW ', 'player': player, 'player_see': str(tile)}]
```
Comment (Ledenel) [inline mahjong/record/utils/event.py]: `'DRAW '` remains a blank. is this meanful?

File: mahjong/record/utils/event.py [hunk 5]
```diff
@@ -112,3 +113,99 @@ def __str__(self):
...
+        if is_somebody_win_game(event):
+            player = attrs['who']
+            loser = attrs['fromWho']
+            score = number_list(event.attrib['ten'])[1]
+            return [{'event_type': 'WIN  ', 'player': player, 'origin': loser}]
+
+        if is_nobody_win_game(event):
+            return [{'event_type': 'EVEN ', 'origin': DRAWN_TYPES[attrs['type']] if 'type' in attrs else ''}]
+
+        return ['UNPARSED']
```
Comment (Ledenel) [inline mahjong/record/utils/event.py]: `'UNPARSED'` is a plain string, which may cause issues when convert to `pandas.DataFrame`. consider return `[]`, and perform a `logging.warn()` to mention something (which?) not recognized.

File: tests/test_paifu.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+class TestTile(TestCase):
+    def test_tile_from_tenhou(self):
+        a = from_file("tests/2009060321gm-00b9-0000-75b25bcf.xml")
+        retlen = []
+        for z in a.game_list:
+            for i in z.events:
+                ret = i.to_paifu()
+                retlen.append(len(ret))
+        assert retlen == [5, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
```
Comment (Ledenel) [inline tests/test_paifu.py]: This assertion assume that `INIT` must have been parsed to 5 events, which is conter-intutive and weird for developer and not flexable. Consider seperate it into another unit test (method), and name the new unit test properly (like `test_INIT_must_have_5_event`)

**Comment by Ledenel** [comment]
Thanks for the contribution!

If there's some unit test prove that `to_paifu()` convert to a real `pandas.DataFrame` is possible, we could be more confident to merge so much code in the codebase. 

you can try to add dependency through 
https://github.com/Ledenel/auto-white-reimu/blob/a9c38069111cba525d1a8c92a53e71e5f2ada99a/setup.py#L21
to make conversion test only affect developers and Travis CI (don't require user to install pandas and huge dependencies)

File: mahjong/record/utils/event.py (hunk 1 above)
Comment (canuse) [reply mahjong/record/utils/event.py]: No special meanings, just to make sure the length of  'event_type' is 5 in all conditions.

**Comment by Ledenel** [comment]
sorry my fault.

https://stackoverflow.com/questions/29870629/pip-install-test-dependencies-for-tox-from-setup-py

Here is a solution to make everything that test needs get properly installed.

File: mahjong/record/utils/event.py (hunk 1 above)
Comment (Ledenel) [reply mahjong/record/utils/event.py]: a `DataFrame` will always treat string as pure python object, so no fixed length needed.

File: .travis.yml [hunk 7]
```diff
@@ -4,6 +4,7 @@ python: 3.7
 before_install:
   - pip install pytest pytest-cov
   - pip install coveralls
+  - pip install pandas
```
Comment (Ledenel) [inline .travis.yml]: For now you could let `pandas` and `pytest` to be together installed by `pip install .[test]`.
you could:
```
pip install .[test]
pip install pytest-cov coveralls
```

to put them together.

**Comment by Ledenel** [comment]
LGTM! Thanks for fixing coverage issues, by the way.

**Pull request merged**